File: app/src/main/java/com/blighter/algoprog/api/LoginMethods.java
package com.blighter.algoprog.api;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.style.ForegroundColorSpan;
import android.util.Log;
import android.util.Pair;
import android.view.Menu;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.room.Room;

import com.blighter.algoprog.R;
import com.blighter.algoprog.data.Achievement;
import com.blighter.algoprog.data.AchievementDao;
import com.blighter.algoprog.data.Database;
import com.blighter.algoprog.pojo.Cookies;
import com.blighter.algoprog.pojo.UserData;
import com.blighter.algoprog.pojo.myUser;
import com.blighter.algoprog.retrofit.AuthorizationInterface;
import com.blighter.algoprog.retrofit.Client;
import com.blighter.algoprog.retrofit.MyUserInterface;
import com.blighter.algoprog.utils.RoundedBackgroundSpan;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.textfield.TextInputLayout;

import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

//this class contains methods which are interact with ActionBar and cookies and could be used in several places
public class LoginMethods {
    //this method using a formula that Kalinin created to generate color for user name
    public static float[] getHSV(int rating, double activity) {
        double R = 160000.0;
        double A = 7.0;
        double h = 11.0 / 12.0 * (1.0 - Math.log(rating + 1.0) / Math.log(R + 1.0));
        double v = 0.3 + 0.7 * Math.log(activity + 1.0) / Math.log(A + 1.0);
        if (0.1 > activity) {
            v = 0;
        }
        h *= 366;
        float h1 = (float) h;
        float v1 = (float) v;
        return new float[]{h1, 100, v1};

    }

    //This method gets new Cookies, then getting user info, building string for ActionBar using got info and setting built strings
    public static CompositeDisposable getCookiesAndMyUser(Context context, final androidx.appcompat.app.ActionBar actionBar, UserData userData, NavigationView navView, Boolean firstTimeLoggingIn, @Nullable TextInputLayout textInputLayout) {
        AuthorizationInterface apiForCookies = Client.getClient().create(AuthorizationInterface.class);
        MyUserInterface apiForMyUser = Client.getClient().create(MyUserInterface.class);
        CompositeDisposable disposables = new CompositeDisposable();
        final boolean[] cookiesOnSuccess = {true};
        //getting cookies using userData
        apiForCookies.getCookies(userData)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .switchMap((Function<Response<Cookies>, Observable<myUser>>) cookiesResponse -> {
                    //if we got wrong userData we return null and set cookiesOnSuccess to false
                    if (!cookiesResponse.isSuccessful()) {
                        cookiesOnSuccess[0] = false;
                        return null;
                    } else {
                        //getting cookies from headers
                        String[] cookiesAndSomething = cookiesResponse.headers().get("Set-Cookie").split(";");
                        SharedPreferences sharedPreferences = context.getSharedPreferences(context.getString(R.string.app_preferences), Context.MODE_PRIVATE);
                        SharedPreferences.Editor editor = sharedPreferences.edit();
                        editor.putString(context.getString(R.string.cookies), cookiesAndSomething[0]).apply();
                        //this is another network request with our cookies to get information about user
                        return apiForMyUser.getMyUserInfo(cookiesAndSomething[0]);
                    }
                })
                .observeOn(Schedulers.io())
                .subscribeOn(Schedulers.io())
                .switchMap(new Function<myUser, Observable<Pair<SpannableString, SpannableStringBuilder>>>() {
                    @Override
                    public Observable<Pair<SpannableString, SpannableStringBuilder>> apply(myUser myUserResponse) throws Exception {
                        //building 2 strings: title and subtitle
                        SpannableStringBuilder coolSubtitle = new SpannableStringBuilder();
                        String name = myUserResponse.getName();
                        String rating = myUserResponse.getRating().toString();
                        double fullActivity = myUserResponse.getActivity();
                        int firstPartOfActivity = (int) fullActivity;
                        int secondPartOfActivity = (int) ((fullActivity * 10) % 10);
                        String activity = firstPartOfActivity + "." + secondPartOfActivity;
                        SpannableString coolSubtitlePartOne = new SpannableString(rating);
                        SpannableString coolSubtitlePartTwo = new SpannableString("/" + activity);
                        coolSubtitlePartOne.setSpan(new ForegroundColorSpan(Color.HSVToColor(getHSV(myUserResponse.getRating(), myUserResponse.getActivity()))), 0, rating.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                        coolSubtitle.append(coolSubtitlePartOne);
                        coolSubtitlePartTwo.setSpan(new ForegroundColorSpan(Color.BLACK), 0, activity.length() + 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                        coolSubtitle.append(coolSubtitlePartTwo);
                        coolSubtitle.append(" ");
                        SpannableString coolTitle = new SpannableString(myUserResponse.getName() + "  " + myUserResponse.getLevel().getCurrent());
                        coolTitle.setSpan(new ForegroundColorSpan(Color.HSVToColor(getHSV(myUserResponse.getRating(), myUserResponse.getActivity()))), 0, name.length(), 0);
                        //getting database instance from assets to to set achievements in subtitle
                        Database db = Room.databaseBuilder(context, Database.class, "achievements.db")
                                .fallbackToDestructiveMigration()
                                .createFromAsset("databases/achievements.db")
                                .build();
                        AchievementDao dao = db.getDao();
                        //adding achievements to subTitle
                        for (int i = 0; i < 3; i++) {
                            Achievement achievement1 = dao.getById(myUserResponse.getAchieves().get(i));
                            SpannableString text = new SpannableString(achievement1.getText());
                            text.setSpan(new RoundedBackgroundSpan(Color.parseColor(achievement1.getColor())), 0, text.length(), 0);
                            coolSubtitle.append(text);
                            coolSubtitle.append(" ");
                        }
                        //if we dont have userId we insert it
                        SharedPreferences sharedPreferences = context.getSharedPreferences(context.getString(R.string.app_preferences), Context.MODE_PRIVATE);
                        if (!sharedPreferences.contains(context.getString(R.string.user_id))) {
                            SharedPreferences.Editor editor = sharedPreferences.edit();
                            editor.putString(context.getString(R.string.user_id), myUserResponse.get_id()).apply();
                        }
                        //creating Observable with 2 string, this made because
                        //we want to set those string on MainThread but build strings on background thread
                        Observable<Pair<SpannableString, SpannableStringBuilder>> strings = Observable
                                .create((ObservableOnSubscribe<Pair<SpannableString, SpannableStringBuilder>>) emitter -> {
                                    if (!emitter.isDisposed()) {
                                        Pair<SpannableString, SpannableStringBuilder> titles = new Pair<>(coolTitle, coolSubtitle);
                                        emitter.onNext(titles);
                                        emitter.onComplete();
                                    }
                                });
                        return strings;
                    }
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Pair<SpannableString, SpannableStringBuilder>>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        disposables.add(d);
                    }

                    @Override
                    public void onNext(Pair<SpannableString, SpannableStringBuilder> titles) {
                        //if we arrived here it means there are no errors occurred and we can set
                        //login and password + the fact that we got new cookies
                        SharedPreferences data = context.getSharedPreferences(context.getString(R.string.app_preferences), Context.MODE_PRIVATE);
                        SharedPreferences.Editor editor = data.edit();
                        editor.putString(context.getString(R.string.login), userData.getUsername());
                        editor.putString(context.getString(R.string.password), userData.getPassword());
                        editor.putBoolean(context.getString(R.string.oldCookies), false);
                        editor.putBoolean(context.getString(R.string.noCookies), false);
                        editor.apply();
                        actionBar.setTitle(titles.first);
                        actionBar.setSubtitle(titles.second);
                        Menu menu = navView.getMenu();
                        menu.clear();
                        navView.inflateMenu(R.menu.drawer_logged);
                        navView.invalidate();
                        if (firstTimeLoggingIn) {
                            ((Activity) context).onBackPressed();
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.d("сука", e.getMessage());
                        SharedPreferences data = context.getSharedPreferences(context.getString(R.string.app_preferences), Context.MODE_PRIVATE);
                        SharedPreferences.Editor editor = data.edit();
                        editor.putBoolean(context.getString(R.string.oldCookies), true);
                        editor.putBoolean(context.getString(R.string.noCookies), true);
                        editor.apply();
                        if (cookiesOnSuccess[0]) {
                            textInputLayout.setError("Неизвестная ошибка");
                        } else {
                            textInputLayout.setError("Неверный логин или пароль");
                        }
                    }

                    @Override
                    public void onComplete() {
                    }
                });
        return disposables;
    }

    //basically the same as getCookiesAndMyUser but we already have cookies and dont do network request to get them
    private static CompositeDisposable getMyUser(String cookies, final androidx.appcompat.app.ActionBar actionBar, Context context) {
        CompositeDisposable disposables = new CompositeDisposable();
        MyUserInterface apiForMyUser = Client.getClient().create(MyUserInterface.class);
        apiForMyUser.getMyUserInfo(cookies)
                .observeOn(Schedulers.io())
                .subscribeOn(Schedulers.io())
                .switchMap((Function<myUser, Observable<Pair<SpannableString, SpannableStringBuilder>>>) myUserResponse -> {
                    SpannableStringBuilder coolSubtitle = new SpannableStringBuilder();
                    String name = myUserResponse.getName();
                    String rating = myUserResponse.getRating().toString();
                    double fullActivity = myUserResponse.getActivity();
                    int firstPartOfActivity = (int) fullActivity;
                    int secondPartOfActivity = (int) ((fullActivity * 10) % 10);
                    String activity = firstPartOfActivity + "." + secondPartOfActivity;
                    SpannableString coolSubtitlePartOne = new SpannableString(rating);
                    SpannableString coolSubtitlePartTwo = new SpannableString("/" + activity);
                    coolSubtitlePartOne.setSpan(new ForegroundColorSpan(Color.HSVToColor(getHSV(myUserResponse.getRating(), myUserResponse.getActivity()))), 0, rating.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                    coolSubtitle.append(coolSubtitlePartOne);
                    coolSubtitlePartTwo.setSpan(new ForegroundColorSpan(Color.BLACK), 0, activity.length() + 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                    coolSubtitle.append(coolSubtitlePartTwo);
                    coolSubtitle.append(" ");
                    SpannableString coolTitle = new SpannableString(myUserResponse.getName() + "  " + myUserResponse.getLevel().getCurrent());
                    coolTitle.setSpan(new ForegroundColorSpan(Color.HSVToColor(getHSV(myUserResponse.getRating(), myUserResponse.getActivity()))), 0, name.length(), 0);
                    Database db = Room.databaseBuilder(context, Database.class, "achievements.db")
                            .fallbackToDestructiveMigration()
                            .createFromAsset("databases/achievements.db")
                            .build();
                    AchievementDao dao = db.getDao();
                    for (int i = 0; i < 3; i++) {
                        Achievement achievement1 = dao.getById(myUserResponse.getAchieves().get(i));
                        SpannableString text = new SpannableString(achievement1.getText());
                        text.setSpan(new RoundedBackgroundSpan(Color.parseColor(achievement1.getColor())), 0, text.length(), 0);
                        coolSubtitle.append(text);
                        coolSubtitle.append(" ");
                    }
                    Observable<Pair<SpannableString, SpannableStringBuilder>> strings = Observable
                            .create((ObservableOnSubscribe<Pair<SpannableString, SpannableStringBuilder>>) emitter -> {
                                if (!emitter.isDisposed()) {
                                    Pair<SpannableString, SpannableStringBuilder> titles = new Pair<>(coolTitle, coolSubtitle);
                                    emitter.onNext(titles);
                                    emitter.onComplete();
                                }
                            })
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread());
                    return strings;
                })
                .subscribe(new Observer<Pair<SpannableString, SpannableStringBuilder>>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        disposables.add(d);
                    }

                    @Override
                    public void onNext(Pair<SpannableString, SpannableStringBuilder> titles) {
                        actionBar.setTitle(titles.first);
                        actionBar.setSubtitle(titles.second);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Toast.makeText(context, "Неизвестная ошибка при обновление информации о пользователе", Toast.LENGTH_LONG).show();
                    }

                    @Override
                    public void onComplete() {

                    }
                });
        return disposables;
    }

    //this method checks if we have new Cookies or have userData and decides which login method to perform
    public static CompositeDisposable setNiceTitle(androidx.appcompat.app.ActionBar actionBar, Context context, NavigationView navigationView) {
        SharedPreferences sharedPref = context.getSharedPreferences(context.getString(R.string.app_preferences), Context.MODE_PRIVATE);
        boolean noCookies = sharedPref.getBoolean(context.getString(R.string.noCookies), true);
        boolean oldCookies = sharedPref.getBoolean(context.getString(R.string.oldCookies), true);
        CompositeDisposable compositeDisposable = new CompositeDisposable();
        if (!noCookies) {
            if (!oldCookies) {
                compositeDisposable = getMyUser(sharedPref.getString(context.getString(R.string.cookies), ""), actionBar, context);
            } else {
                SharedPreferences.Editor editor = sharedPref.edit();
                editor.putBoolean(context.getString(R.string.oldCookies), false);
                editor.apply();
                compositeDisposable = getCookiesAndMyUser(context, actionBar, new UserData(sharedPref.getString(context.getString(R.string.login), ""), sharedPref.getString(context.getString(R.string.password), "")), navigationView, false, null);
            }
        } else {
            actionBar.setTitle("Неизвестный пользователь");
        }
        return compositeDisposable;
    }
}

File: app/src/main/java/com/blighter/algoprog/api/LoginMethods.kt
package com.blighter.algoprog.api

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.style.ForegroundColorSpan
import android.widget.Toast
import androidx.appcompat.app.ActionBar
import androidx.room.Room
import com.blighter.algoprog.R
import com.blighter.algoprog.data.Database
import com.blighter.algoprog.network.*
import com.blighter.algoprog.network.ClientCoroutines.client
import com.blighter.algoprog.utils.RoundedBackgroundSpan
import com.google.android.material.navigation.NavigationView
import kotlinx.coroutines.*
import retrofit2.Response
import kotlin.math.ln

//this method using a formula that Kalinin created to generate color for user name
fun getHSV(rating: Int, activity: Double): FloatArray? {
    val r = 160000.0
    val a = 7.0
    var h = 11.0 / 12.0 * (1.0 - ln(rating + 1.0) / ln(r + 1.0))
    var v = 0.3 + 0.7 * ln(activity + 1.0) / ln(a + 1.0)
    if (0.1 > activity) {
        v = 0.0
    }
    h *= 366.0
    val h1 = h.toFloat()
    val v1 = v.toFloat()
    return floatArrayOf(h1, 100f, v1)
}

sealed class Errors(val warn: String)
class Unknown : Errors("Неизвестная ошибка")
class NoInternet : Errors("Нет подключения к интернету")
class WrongLogPass : Errors("Неверный логин или пароль")


data class MyUserInfoResponse(val error: String?, val title: SpannableString?, val subtitle: SpannableStringBuilder?, val myUser: MyUser?)

suspend fun getMyUserInfoAndCookies(userData: UserData?, context: Context): Deferred<MyUserInfoResponse?> = coroutineScope {
    async {
        var myUserInfoResponse: MyUserInfoResponse? = null
        var cookiesResponse: Response<Cookies?>? = null
        var error: String? = null
        try {
            cookiesResponse = client.create(AuthorizationInterface::class.java).getCookies(userData)?.execute()
        } catch (e: Exception) {
            error = NoInternet().warn
        }
        if (error != null) {
            myUserInfoResponse = MyUserInfoResponse(error, null, null, null)
        } else if (!cookiesResponse!!.isSuccessful) {
            myUserInfoResponse = MyUserInfoResponse(WrongLogPass().warn, null, null, null)
        } else {
            val cookies = cookiesResponse.headers().toString()
                    .substringAfter("set-cookie:")
                    .takeWhile { it != ';' }
//saving it in sharPref
            val sharedPreferences = context.getSharedPreferences("justSomeData", Context.MODE_PRIVATE)
            val editor = sharedPreferences.edit()
            editor.putString(context.getString(R.string.cookies), cookies).apply()
            myUserInfoResponse = getMyUserInfo(cookies, context).await()
//if we dont have userId we insert it
            if (!sharedPreferences.contains(context.getString(R.string.user_id))) {
                editor.putString(context.getString(R.string.user_id), myUserInfoResponse?.myUser?._id).apply()
            }
            editor.putString(context.getString(R.string.login), userData?.username)
            editor.putString(context.getString(R.string.password), userData?.password)
            editor.putBoolean(context.getString(R.string.oldCookies), false)
            editor.putBoolean(context.getString(R.string.noCookies), false)
            editor.apply()
        }
        myUserInfoResponse
    }
}

suspend fun getMyUserInfo(cookies: String?, context: Context): Deferred<MyUserInfoResponse?> = coroutineScope {
    async {
        var myUser: MyUser? = null
        var error: Boolean? = null
        var myUserResponse: MyUserInfoResponse? = null
        try {
            myUser = client.create(MyUserInterface::class.java).getMyUserInfo(cookies)?.execute()?.body()
        } catch (e: Exception) {
            error = true
        }

        when {
            error == true -> {
                myUserResponse = MyUserInfoResponse(NoInternet().warn, null, null, null)
            }
            myUser == null -> {
                myUserResponse = MyUserInfoResponse(Unknown().warn, null, null, null)
            }
            else -> {
                val coolSubtitle = SpannableStringBuilder()
                val firstPartOfActivity = myUser.activity.toInt()
                val secondPartOfActivity = (myUser.activity * 10 % 10).toInt()
                val activity = "$firstPartOfActivity.$secondPartOfActivity"
                val rating = myUser.rating.toString()
                val coolSubtitlePartOne =

                        SpannableString(rating)
                val coolSubtitlePartTwo = SpannableString("/$activity")
                coolSubtitlePartOne.setSpan(ForegroundColorSpan
                (Color.HSVToColor(
                        getHSV(myUser.rating, myUser.activity))), 0, rating.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                coolSubtitle.append(coolSubtitlePartOne)
                coolSubtitlePartTwo.setSpan(ForegroundColorSpan(Color.BLACK), 0, activity.length + 1, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                coolSubtitle.append(coolSubtitlePartTwo)
                coolSubtitle.append(" ")
                val coolTitle = SpannableString(myUser.name + " " + myUser.level.current)
                coolTitle.setSpan(ForegroundColorSpan(
                        Color.HSVToColor(
                                getHSV(myUser.rating, myUser.activity))), 0, myUser.name.length, 0)
                val db = Room.databaseBuilder(context, Database::class.java, "achievements.db")
                        .fallbackToDestructiveMigration()
                        .createFromAsset("databases/achievements.db")
                        .build()
                val dao = db.getDao()
                for (i in 0..2) {
                    val achievement = dao.getById(myUser.achieves[i])
                    val text = SpannableString(achievement!!.text)
                    text.setSpan(RoundedBackgroundSpan(Color.parseColor(achievement.color)), 0, text.length, 0)
                    coolSubtitle.append(text)
                    coolSubtitle.append(" ")
                }
                myUserResponse = MyUserInfoResponse(null, coolTitle, coolSubtitle, myUser)
            }
        }
        myUserResponse
    }
}

fun setNiceTitle(actionBar: ActionBar?, context: Context, navView: NavigationView, firstTimeLoggingIn: Boolean, userData: UserData?) {
    val sharedPref = context.getSharedPreferences(context.getString(R.string.app_preferences), Context.MODE_PRIVATE)
    val noCookies = sharedPref.getBoolean(context.getString(R.string.noCookies), true)
    val oldCookies = sharedPref.getBoolean(context.getString(R.string.oldCookies), true)
    var myUserInfoResponse: MyUserInfoResponse? = null
    if (!noCookies) {
        scopeNetwork.launch {
            if (!oldCookies) {
                myUserInfoResponse = getMyUserInfo(sharedPref.getString(context.getString(R.string.cookies), ""), context).await()
            } else {
                val editor = sharedPref.edit()
                //TODO maybe we should say that cookies aren't old if we sure that we got them and they are good
                editor.putBoolean(context.getString(R.string.oldCookies), false).apply()
                if (userData == null) {
                    if (sharedPref.getString(context.getString(R.string.login), "") != "") {
                        val userDataFromPref = UserData(sharedPref.getString(context.getString(R.string.login), ""), sharedPref.getString(context.getString(R.string.password), ""))
                        myUserInfoResponse = getMyUserInfoAndCookies(userDataFromPref, context).await()
                    }
                } else {
                    myUserInfoResponse = getMyUserInfoAndCookies(userData, context).await()
                }
            }
            withContext(Dispatchers.Main) {
                if (myUserInfoResponse != null && myUserInfoResponse?.error == null) {
                    actionBar?.title = myUserInfoResponse!!.title
                    actionBar?.subtitle = myUserInfoResponse!!.subtitle
                    val menu = navView.menu
                    menu.clear()
                    navView.inflateMenu(R.menu.drawer_logged)
                    navView.invalidate()
                    if (firstTimeLoggingIn) {
                        (context as Activity).onBackPressed()
                    }
                } else {
                    if (myUserInfoResponse?.error != null) {
                        Toast.makeText(context, myUserInfoResponse?.error, Toast.LENGTH_LONG).show()
                    }
                }
            }
        }
    } else {
        actionBar?.setTitle("Неизвестный пользователь")
    }
}